Set Home Page Preference

dtaylor113 opened the pull request:
![image](https://user-images.githubusercontent.com/12733153/34109527-dd6ae042-e3d1-11e7-9047-405cee01be55.png)

![image](https://user-images.githubusercontent.com/12733153/34109549-ef029bf6-e3d1-11e7-9f3e-a6dd424cd71f.png)

Trello Card: https://trello.com/c/sg1xw7gG

spadgett commented on app/scripts/services/homePagePreference.js:
```diff
@@ -0,0 +1,105 @@
+      return _.get(homePagePref, 'type', 'catalog-home');
+    };
+
+    var getProjectOverviewPage = function() {
```
nit: It's getting a project name, not an overview page (which to me implies a URL). Maybe

```js
var getHomePageProjectName = function() ...
```

spadgett commented on app/scripts/services/homePagePreference.js:
```diff
@@ -0,0 +1,105 @@
+      });
+    };
+
+    var getHomePageURL = function() {
```
nit: getHomePagePath is more precise

spadgett commented on app/scripts/services/homePagePreference.js:
```diff
@@ -0,0 +1,105 @@
+
+    var setHomePagePreference = function(homePagePref) {
+      var key = getHomePagePreferenceKey();
+      if (!key) {
```
`key` is always set

spadgett commented on app/scripts/services/homePagePreference.js:
```diff
@@ -0,0 +1,105 @@
+      setHomePagePreference: setHomePagePreference,
+      getProjectOverviewPage: getProjectOverviewPage,
+      getHomePageURL: getHomePageURL,
+      notifyInvalidProjectHomePage: notifyInvalidProjectHomePage,
```
trailing comma

spadgett commented on app/styles/_set-home-page.less:
```diff
@@ -0,0 +1,10 @@
+.modal-body-indent {
```
These classes are very specific to this dialog and should be scoped to just the home page modal.

spadgett commented on app/views/modals/set-home-page-modal.html:
```diff
@@ -0,0 +1,50 @@
+      <div class="title">Set your preferred home page.</div>
+      Note: This setting is browser specific and will not be maintained across browsers.
+    </p>
+    <ng-form name="homePageForm">
```
`<form>` instead of `<ng-form>`. `ng-form` should really only be used if it's nested inside another `form` element.

spadgett commented on app/views/modals/set-home-page-modal.html:
```diff
@@ -0,0 +1,50 @@
+        <div class="radio" ng-show="availableProjects.length >= 1">
+          <label>
+            <input type="radio" ng-model="preferredHomePage" value="project-overview">
+            Overview Page for {{availableProjects.length === 1 ? (selectedProject | displayName) : ':'}}
```
this will put an extra space in front of the `:`

spadgett commented on app/views/modals/set-home-page-modal.html:
```diff
@@ -0,0 +1,50 @@
+            Overview Page for {{availableProjects.length === 1 ? (selectedProject | displayName) : ':'}}
+          </label>
+          <div class="select-project-container">
+            <select-project ng-if="availableProjects.length > 1"
```
nit: if you move `ng-if` to the parent div it avoids the empty div

spadgett commented on app/views/modals/set-home-page-modal.html:
```diff
@@ -0,0 +1,50 @@
+          </label>
+          <div class="select-project-container">
+            <select-project ng-if="availableProjects.length > 1"
+                            class="select-project"
```
Does this class have any styles? I don't see it in this PR.

dtaylor113 replied:
Removed, not needed

spadgett commented on app/views/modals/set-home-page-modal.html:
```diff
@@ -0,0 +1,50 @@
+  </div>
+  <div class="modal-footer">
+      <button class="btn btn-lg btn-primary" type="button"
+              ng-disabled="(preferredHomePage === 'project-overview' && homePageForm.$invalid) || homePageForm.$pristine"
```
I think this can just be... ?

```html
ng-disabled="homePageForm.$invalid || homePageForm.$pristine"
```

spadgett replied:
Nevermind I see why it's not possible.

It would be better to add an `ng-required` attribute to the `select-project` component and then use...

```html
<select-project
  ng-required="preferredHomePage === 'project-overview'"
  ....
```

Then you can change this disabled check to the above.

dtaylor113 replied:
Thanks for the tip, I hadn't used `ng-required` before...very useful :-)

dtaylor113 commented on app/scripts/app.js:
```diff
@@ -67,7 +72,7 @@ angular
       .when('/project/:project', {
         redirectTo: function(params) {
-          return '/project/' + encodeURIComponent(params.project) + "/overview";
+          return '/project/' + encodeURIComponent(params.project) + "/overview" + (params.isHomePage ? "?isHomePage=true" : "");
```
Removed, it was not needed  -thanks

dtaylor113 commented:
Hi @spadgett, thanks for the great review.  I believe I have addressed your last round of comments.

spadgett commented:
@openshift/team-ux-review Since we've changed clicking on the logo to always go to the landing page, there is no direct link to your preferred home page in the UI. You literally have to type in "/console/" as the URL in the browser location bar. Should we have a way to navigate to your home page somewhere in the UI?

cc @jwforres

dtaylor113 commented:
> @openshift/team-ux-review Since we've changed clicking on the logo to always go to the landing page, there is no direct link to your preferred home page in the UI. You literally have to type in "/console/" as the URL in the browser location bar. Should we have a way to navigate to your home page somewhere in the UI?

Good question, cc'ing @beanh66

dtaylor113 commented:
Hi, I also just noticed:

![image](https://user-images.githubusercontent.com/12733153/34575046-5434212a-f147-11e7-9e7c-58aa4b2c1e5a.png)

Since home page is now configurable to places other than the Catalog, I think this Help menu option should change to something like 'Tour Catalog".   I also noticed that selecting `Help -> Tour Home Page` doesn't do anything if you are on a page other than the Browse Catalog landing page, but I think I can fix that.

spadgett commented:
>  I also noticed that selecting Help -> Tour Home Page doesn't do anything if you are on a page other than the Browse Catalog landing page, but I think I can fix that.

Yeah seems to work in master, so it is due to the redirect. We should probably sweep the code for anywhere else we are linking to `./`

spadgett commented:
All of these need to be updated.

```
app/scripts/services/navigate.js
94:          return './';

app/scripts/controllers/membership.js
306:                      $location.url("./");

app/views/overview/_service-bindings.html
22:      <a href="./">Browse Catalog</a>

app/scripts/extensions/nav/helpDropdown.js
23:                node: '<li><a href="./?startTour=true">Tour Home Page</a></li>'
```

dtaylor113 commented:
Updated those references to `./`. Also changed `Help -> Tour Home Page` to:

![image](https://user-images.githubusercontent.com/12733153/34581811-c75c04f2-f15f-11e7-8591-31e26baa1500.png)

Since that's what we call it in the Set Home Page modal:
![image](https://user-images.githubusercontent.com/12733153/34581825-d304e2b0-f15f-11e7-8bde-2a98f90d08bd.png)

ncameronbritt commented:
@spadgett The designers chatted about the need for another way to get to the homepage and we feel like things are ok as is. Our feeling is that if the user has changed their homepage, they can still to navigate to whatever page they selected within a couple of clicks, regardless of where they are in the console.

spadgett commented:
> The designers chatted about the need for another way to get to the homepage and we feel like things are ok as is. Our feeling is that if the user has changed their homepage, they can still to navigate to whatever page they selected within a couple of clicks, regardless of where they are in the console.

It's not so much about getting back to your homepage. My question is how you get to your homepage *at all*?

Presumably someone sends you a link to the console the first time you use it. Homepage is implemented as a redirect currently, so you will be redirected to the landing page, the default homepage, after logging in the first time. The URL changes to `/console/catalog`. If you create a bookmark now, it will always go to the catalog even if you change your homepage. You're bookmarking `/console/catalog`, the catalog page, and not `/console`, your homepage. You have no way to get back to your selected homepage without finding that first link someone sent you.

If we want this to work differently, we can't use redirects. We might need to revisit our implementation (sorry @dtaylor113 this would be my fault).

I guess my question is still if you have a browser bookmark anyway, why not just bookmark the page you want?

dtaylor113 commented:
> We might need to revisit our implementation

@spadgett, would it be possible to do this work in a follow up PR?

spadgett commented:
> would it be possible to do this work in a follow up PR?

I think even a link that bookmarks your homepage or goes to your homepage would make it more useful without needing change the implementation.

dtaylor113 commented:
> I think even a link that bookmarks your homepage or goes to your homepage would make it more useful without needing change the implementation.

Do you mean updating app.js to include something like `/my-home-page', which would do the same as  `/` does, ie. `return HomePagePreferenceServiceProvider.$get().getHomePagePath();`?

dtaylor113 commented:
I'm working on getting the integration tests to work on my Linux machine to debug the issue

spadgett commented:
@dtaylor113 can you check create from URL with your changes?

https://docs.openshift.org/latest/dev_guide/create_from_url.html

dtaylor113 commented:
> @dtaylor113 can you check create from URL with your changes?

Ok, will try that.

> Jenkins/Integration Tests failure. 

I'm seeing this failure on my system:
```
visiting url: create-project
visiting url: project/template-dumpster/create

/home/dtaylor/Dev/repos/openshift/origin-web-console/node_modules/selenium-webdriver/lib/webdriver/promise.js:1654
      var result = fn();
                   ^
Error: URL hasn't changed to project/template-dumpster/create(is currently: undefined)
Wait timed out after 5190ms
    at /home/dtaylor/Dev/repos/openshift/origin-web-console/node_modules/selenium-webdriver/lib/webdriver/promise.js:1425:29
```

dtaylor113 commented:
> @dtaylor113 can you check create from URL with your changes?
> 
> https://docs.openshift.org/latest/dev_guide/create_from_url.html
> 
> Those are the tests that are failing.

I used `https://localhost:9000/dev-console/create?imageStream=nodejs&imageTag=4&name=nodejs&sourceURI=https:%2F%2Fgithub.com%2Fopenshift%2Fnodejs-ex.git&sourceRef=master&contextDir=%2F` and go to this point:

![image](https://user-images.githubusercontent.com/12733153/34730214-b8707902-f52c-11e7-8f25-ac9ce052c5a3.png)

`the API version in the data (v1) does not match the expected API version (project.openshift.io/v1)` ?

dtaylor113 commented:
@spadgett, @benjaminapetersen,  PR passed Jenkis and Travis tests.  Also the 'Create From URL' test worked in local testing.

spadgett commented:
/lgtm

spadgett commented:
Thanks @dtaylor113
